### ISSUE TITLE
How to run Crazy Craft?

### ISSUE THREAD

**Issue by nayneyT** [itzg/docker-minecraft-server#703]
Hi,

Is there any way of running crazy craft 3.0 or 4.0 using this docker container?

Many thanks,

Tom

**Comment by itzg** [comment]
Yes, but you'll need to pull the latest images since my testing of this scenario uncovered a couple missing bits.

Download the server pack [from here](http://voidswrath.com/modpacks/crazy-craft-3-0/)

Create a new directory to hold the following `docker-compose.yml` file, as an example:
```yaml
version: "3"

services: 
  mc:
    image: itzg/minecraft-server
    environment: 
      EULA: "TRUE"
      TYPE: CUSTOM
      GENERIC_PACK: /modpacks/CrazyCraft 3.zip
      CUSTOM_SERVER: /data/crazycraft.jar
      MEMORY: 2G
    volumes: 
      - ./modpacks:/modpacks:ro
      - crazycraft:/data
    ports: 
      - 25565:25565

volumes: 
  crazycraft: {}
```

Place the downloaded server pack zip in a new subdirectory called `modpacks`, as shown here:

![image](https://user-images.githubusercontent.com/988985/103300713-24388180-49c5-11eb-8cbf-a9bd830e4a08.png)

When you bring that up, it'll unzip the downloaded server modpack and run the included crazycraft.jar server file.

**Comment by nayneyT** [comment]
Hi,

I apologise for the length of time it has taken me to replying.

This this the docker-compose file I have tried:

```
version: '3'

services:
  minecraft-server-crazy:
    image: itzg/minecraft-server:latest
    container_name: minecraft-server-crazy
    hostname: minecraft-server-crazy

    ports:
      - "25565:25565"

    environment:
      EULA: "TRUE"
      TYPE: CUSTOM
      GENERIC_PACK: /modpacks/CrazyCraft 3.zip
      CUSTOM_SERVER: /data/crazycraft.jar
      MEMORY: "2G" # "1G" by default, can be used to adjust both initial (Xms) and max (Xmx) memory settings of the JVM
      SERVER_PORT: 25565

    restart: always
    
    volumes:
      - ./modpacks:/modpacks:ro
      - crazycraft:/data

volumes: 
  crazycraft: {}
```

The folder layout is as follows:

![mine1](https://user-images.githubusercontent.com/37027153/106390385-f92bbe00-63c6-11eb-8396-3639f59abc01.png)
![mine2](https://user-images.githubusercontent.com/37027153/106390386-f92bbe00-63c6-11eb-9119-ce859e4efd39.png)

Unfortunately I am getting the following error from Docker Logs

```
[init] Running as uid=1000 gid=1000 with /data as 'drwxrwxrwx    2 1000     1000          4096 Jan 31 16:12 /data'
[init] Resolved version given LATEST into 1.16.5
[init] Resolving type given CUSTOM
[init] CUSTOM_SERVER is not properly set to a URL or existing jar file
```

Any assistance would be greatly appreciated. The kids are driving me insane.....

**Comment by go0ro0** [comment]
I have this working although I did it by cloning and modifying the docker image I already had running with Portainer.

**Comment by itzg** [comment]
@nayneyT it looks like you haven't re-pulled the latest image, because yours is not showing the new log message I added:

```
Using custom server jar from generic pack at ...
```

**Comment by nayneyT** [comment]
Hi itzg. I had it set to pull but it wasn't removed the existing image and run docker-compose again and now it works. You are a star. Thank you.

```
version: '3'

services:
  minecraft-server-crazy3:
    image: itzg/minecraft-server:latest
    container_name: minecraft-server-crazy3
    hostname: minecraft-server-crazy
    ports:
      - "25566:25565"
    environment:
      EULA: "TRUE"

###### Custom Server
      TYPE: "CUSTOM"
      GENERIC_PACK: "/modpacks/CrazyCraft 3.zip"
      CUSTOM_SERVER: "crazycraft.jar"

###### JAVA
      MEMORY: "2G" # "1G" by default, can be used to adjust both initial (Xms) and max (Xmx) memory settings of the JVM
      MAX_MEMORY: "8G" # independently sets the max heap size

###### GENERAL MINECRAFT SETTINGS
      OVERRIDE_SERVER_PROPERTIES: 'true' # If you would like to override the server configuration each time the container starts up, you can set the OVERRIDE_SERVER_PROPERTIES environment variable
      ANNOUNCE_PLAYER_ACHIEVEMENTS: 'true' # Allows server to announce when a player gets an achievement.
      SPAWN_ANIMALS: 'true' # Determines if animals will be able to spawn.
      SPAWN_MONSTERS: 'true' # Determines if monsters will be spawned.
      SPAWN_NPCS: 'true' # Determines if villagers will be spawned.
     
      # COLOURS https://wiki.ess3.net/mc/ and https://minecraft.tools/en/color-code.php
      MOTD: 'A \u00A79CrazyCraft 3 \u00A78minecraft server!!!' # The message of the day, shown below each server entry in the UI

      MAX_PLAYERS: 10
      CONSOLE: 'true'
      LEVEL: nayney # the name of the world - also the folder name it creates You can either switch between world saves or run multiple containers with different saves by using the LEVEL option, where the default is "world"

      SERVER_PORT: 25565
      ONLINE_MODE: 'true' # By default, server checks connecting players against Minecraft's account database. If you want to create an offline server or your server is not connected to the internet, you can disable the server to try connecting to minecraft.net to authenticate players 

####### OPS/WHITELIST
      #WHITELIST:  # To whitelist players for your Minecraft server, If the WHITELIST environment variable is not used, any user can join your Minecraft server if it's publicly accessible
      #OPS:  # to add more "op" (aka adminstrator) users to your Minecraft server
      ICON: 'https://www.saga.co.uk/contentlibrary/saga/publishing/verticals/home-and-garden/gardening/garden-wildlife/david-chapman/birds/bluetit768.jpg'

    restart: always
    
    volumes:
      - "../modpacks:/modpacks:ro"
      - "../data:/data"
      - "../plugins:/plugins" # Any files in this filesystem will be copied over to the main /data/plugins filesystem before starting Minecraft
      - "../mods:/mods" # Any files in either of these filesystems will be copied over to the main /data filesystem before starting Minecraft
      - "../config:/config" # Any files in either of these filesystems will be copied over to the main /data filesystem before starting Minecraft

# https://minecraft.gamepedia.com/Server.properties
# https://minecraft.gamepedia.com/Commands
```

**Comment by voruti** [comment]
Close this issue?

**Comment by evasil** [comment]
Hello,
regretfully my docker failing to start with mentioned setup.
Using latest images, tested with both Crazycraft 3 and 4.

The error is below:

```
[init] Resolved version given LATEST into 1.17.1
[init] Resolving type given CUSTOM
[init] Manifest exists from older generic pack, cleaning up ...
[init] Writing generic pack manifest ... 
[init] Applying generic pack ...
[init] server.properties already created, skipping
[init] Checking for JSON files.
[init] Setting initial memory to 2G and max to 8G
[init] Starting the Minecraft server...
A problem occurred running the Server launcher.java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at cpw.mods.fml.relauncher.ServerLaunchWrapper.run(ServerLaunchWrapper.java:43)
	at cpw.mods.fml.relauncher.ServerLaunchWrapper.main(ServerLaunchWrapper.java:12)
Caused by: java.lang.ClassCastException: class jdk.internal.loader.ClassLoaders$AppClassLoader cannot be cast to class java.net.URLClassLoader (jdk.internal.loader.ClassLoaders$AppClassLoader and java.net.URLClassLoader are in module java.base of loader 'bootstrap')
	at net.minecraft.launchwrapper.Launch.<init>(Launch.java:34)
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
	... 6 more
2021-08-15T00:06:01.421+1000	WARN	mc-server-runner	sub-process failed	{"exitCode": 1}
2021-08-15T00:06:01.421+1000	INFO	mc-server-runner	Done

```

**Comment by evasil** [comment]
When using java8 image there is different error:


```
[init] Running as uid=1000 gid=1000 with /data as 'drwxrwsr-x   10 1000     1000          4096 Aug 15 00:11 /data'
[init] Resolved version given LATEST into 1.17.1
[init] Resolving type given CUSTOM
[init] server.properties already created, skipping
[init] Checking for JSON files.
[init] Setting initial memory to 2G and max to 8G
[init] Starting the Minecraft server...
We appear to be missing one or more essential library files.
You will need to add them to your server before FML and Forge will run successfully.java.lang.ClassNotFoundException: net.minecraft.launchwrapper.Launch
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at cpw.mods.fml.relauncher.ServerLaunchWrapper.run(ServerLaunchWrapper.java:25)
	at cpw.mods.fml.relauncher.ServerLaunchWrapper.main(ServerLaunchWrapper.java:12)

```

**Comment by evasil** [comment]
Update: with java8 CrazyCraft 4 seems to start ok.
